iptables/config/config.go

```diff
@@ -27,51 +27,120 @@ type DNS struct {
 type Redirect struct {
 	// NamePrefix is a prefix which will be used go generate chains name
 	NamePrefix string
-	Inbound    *TrafficFlow
-	Outbound   *TrafficFlow
-	DNS        *DNS
+	Inbound    TrafficFlow
+	Outbound   TrafficFlow
+	DNS        DNS
 }
 
 type Chain struct {
 	Name string
 }
 
-func (c *Chain) GetFullName(prefix string) string {
+func (c Chain) GetFullName(prefix string) string {
 	return prefix + c.Name
 }
 
 type Config struct {
-	Owner    *Owner
-	Redirect *Redirect
-	// Output is the place where if any produced informational data (i.e. dump of the rules
-	// which will be applied for helping user to potentially debug if something would
-	// go wrong) will be placed (os.Stdout by default)
-	Output io.Writer
-	// Verbose when set will generate iptables configuration with longer argument/flag names,
-	// additional comments etc.
+	Owner    Owner
+	Redirect Redirect
+	// RuntimeOutput is the place where Any debugging, runtime information
+	// will be placed (os.Stdout by default)
+	RuntimeOutput io.Writer
+	// Verbose when set will generate iptables configuration with longer
+	// argument/flag names, additional comments etc.
 	Verbose bool
 }
 
-func DefaultConfig() *Config {
-	return &Config{
-		Owner: &Owner{UID: "5678", GID: "5678"},
-		Redirect: &Redirect{
+func defaultConfig() Config {
+	return Config{
+		Owner: Owner{UID: "5678", GID: "5678"},
+		Redirect: Redirect{
 			NamePrefix: "",
-			Inbound: &TrafficFlow{
+			Inbound: TrafficFlow{
 				Port:          15006,
-				Chain:         &Chain{Name: "MESH_INBOUND"},
-				RedirectChain: &Chain{Name: "MESH_INBOUND_REDIRECT"},
+				Chain:         Chain{Name: "MESH_INBOUND"},
+				RedirectChain: Chain{Name: "MESH_INBOUND_REDIRECT"},
 				ExcludePorts:  []uint16{},
 			},
-			Outbound: &TrafficFlow{
+			Outbound: TrafficFlow{
 				Port:          15001,
-				Chain:         &Chain{Name: "MESH_OUTBOUND"},
-				RedirectChain: &Chain{Name: "MESH_OUTBOUND_REDIRECT"},
+				Chain:         Chain{Name: "MESH_OUTBOUND"},
+				RedirectChain: Chain{Name: "MESH_OUTBOUND_REDIRECT"},
 				ExcludePorts:  []uint16{},
 			},
-			DNS: &DNS{Port: 15053, Enabled: false, ConntrackZoneSplit: false},
+			DNS: DNS{Port: 15053, Enabled: false, ConntrackZoneSplit: false},
 		},
-		Output:  os.Stdout,
-		Verbose: true,
+		RuntimeOutput: os.Stdout,
+		Verbose:       true,
 	}
 }
+
+func MergeConfigWithDefaults(cfg Config) Config {
```

Comment:
nitpick: Is there a more concise way of doing this? A quick internet search only shows JSON unmarshaling custom parts into the default struct which is not great...

Reply:
I don't know about any good way of doing it, but maybe if the config will grow in the future, we could rethink the approach